### ISSUE TITLE
Question about attention_mask usage in GRPO training

### ISSUE THREAD

**Issue by Ryan-gsq** [Red-Hat-AI-Innovation-Team/async-grpo#4]
I noticed that in grpo_loss.py, when processing mini-batches, all question-answer pairs appear to be flattened into a single sequence, but the attention_mask parameter is passed directly to the model without modification:

outputs = model.model(
    input_ids=input_ids,
    attention_mask=attention_mask,  # This could be None or unmodified
    position_ids=position_ids,
    # ...other parameters
)
Questions
Is passing attention_mask without modification intentional or potentially an oversight?
If multiple samples are concatenated into a single sequence, shouldn't the attention_mask be constructed to prevent cross-sample attention?
Without a properly constructed attention mask, wouldn't this allow the model to attend to tokens from other samples in the batch, potentially causing information leakage?
Additional Context
In the fused_log_probs_and_entropy_fn function, there's an assertion that labels.shape[0] == 1, suggesting the implementation expects a batch size of 1 but with packed sequences. This makes me wonder if the attention mask handling is aligned with this expectation.

Thank you for any clarification on this design decision!

**Comment by aldopareja** [comment]
This is a flash attention trick that enables padding free training. Look it up online! It's the same as the one you're pointing to :)

**Comment by Ryan-gsq** [comment]
i found this , thanks u replay

https://huggingface.co/blog/packing-with-FA2
